File: vscode/src/core/NormalModePolicy.ts
import { ImeAction, VimMode } from './types';

export function isNormalLikeMode(mode: VimMode, hasSelection = false): boolean {
  if (hasSelection) return true;
  return mode !== VimMode.INSERT && mode !== VimMode.UNKNOWN;
}

export function isStrictNormalMode(mode: VimMode, hasSelection = false): boolean {
  return !hasSelection && mode === VimMode.NORMAL;
}

export function resolveNormalModeAction(
  mode: VimMode,
  hasSelection: boolean,
  defaultApplied: boolean,
): ImeAction | null {
  if (!isNormalLikeMode(mode, hasSelection)) return null;
  if (isStrictNormalMode(mode, hasSelection) || !defaultApplied) {
    return ImeAction.ENGLISH;
  }
  return ImeAction.UNCHANGED;
}

export function shouldEnforceNormalEnglish(
  mode: VimMode,
  hasSelection: boolean,
  asciiMode: boolean,
  capsLock: boolean,
): boolean {
  return isStrictNormalMode(mode, hasSelection) && (!asciiMode || capsLock);
}

File: core/src/main/kotlin/com/auto_switch_ime/core/NormalModePolicy.kt
package com.auto_switch_ime.core

object NormalModePolicy {
    fun resolveAction(
        normalLike: Boolean,
        strictNormal: Boolean,
        defaultApplied: Boolean
    ): ImeAction? {
        if (!normalLike) return null
        if (strictNormal || !defaultApplied) return ImeAction.ENGLISH
        return ImeAction.UNCHANGED
    }

    fun shouldEnforceEnglish(strictNormal: Boolean, asciiMode: Boolean, capsLock: Boolean): Boolean {
        return strictNormal && (!asciiMode || capsLock)
    }
}
